include/evaluators/blas2_tree_evaluator.hpp

```diff
@@ -0,0 +1,482 @@
+    l.eval_subexpr_if_needed(NULL, dev);
+    r1.eval_subexpr_if_needed(&l.data(), dev);
+    r2.eval_subexpr_if_needed(&l.data(), dev);
+    r3.eval_subexpr_if_needed(&l.data(), dev);
+    return true;
+  }
+
+  value_type eval(size_t i) {
+    auto dim = r2.getSize();
+
+    auto val =
+        functor_traits<iniAddOp1_struct, value_type, Device>::eval(r2.eval(0));
+    for (size_t j = 0; j < dim; j++) {
+      val += r1.eval(i, j) * r2.eval(j);
+    }
+    l.eval(i) += scl * val;
+    return val;
+  }
+
+  template <typename sharedT>
+  value_type eval(sharedT scratch, cl::sycl::nd_item<1> ndItem) {
+    size_t localid = ndItem.get_local(0);
+    size_t localSz = ndItem.get_local_range(0);
+    size_t groupid = ndItem.get_group(0);
+
+    size_t dimR = r1.getSizeR();
+    size_t dimC = r1.getSizeC();
+
+    size_t rowSz = (localSz / nThr);  // number of rows per each workgroup
+    size_t rowid = groupid * rowSz + localid % rowSz;  // rowid of the thread
+
+    size_t colid = localid / rowSz;  // first column on which thread works
+
+    // Local computations
+    auto val =
+        functor_traits<iniAddOp1_struct, value_type, Device>::eval(r2.eval(0));
+    if (rowid < dimR) {
+      for (size_t j = colid; j < dimC; j += nThr) {
+        val += r1.eval(rowid, j) * r2.eval(j);
+      }
+    }
+
+    scratch[localid] = val;
+    // This barrier is mandatory to be sure the data is on the shared memory
+    ndItem.barrier(cl::sycl::access::fence_space::local_space);
+
+    // Reduction inside the block
+    for (size_t offset = nThr >> 1; offset > 0; offset >>= 1) {
+      if ((rowid < dimR) && (colid < offset)) {
+        scratch[localid] += scratch[localid + offset * rowSz];
+      }
+      // This barrier is mandatory to be sure the data are on the shared memory
+      ndItem.barrier(cl::sycl::access::fence_space::local_space);
+    }
+    // The result is stored in lhs
+    if ((rowid < dimR) && (colid == 0)) {
+      l.eval(rowid) = scl * scratch[localid] + r3.eval(rowid);
+    }
+    return val;
+  }
+};
+
+/*! PrdRowMatCvtMultShm.
+ * @brief TWO KERNELS DOT PRODUCT GEMV
+ * FIRST KERNEL: THE LOCAL COMPUTATIONS ARE MADE
+ * The common data are copied to the scratch vector,
+ * and later the computation begins.
+ */
+template <class LHS, class RHS1, class RHS2, typename Device>
+struct Evaluator<PrdRowMatVctMultShmExpr<LHS, RHS1, RHS2>, Device> {
+  using Expression = PrdRowMatVctMultShmExpr<LHS, RHS1, RHS2>;
+  using value_type = typename Expression::value_type;
+  using cont_type = typename Evaluator<LHS, Device>::cont_type;
+  static constexpr bool supported = LHS::supported && RHS1::supported && RHS2::supported;
+
+  Evaluator<LHS, Device> l;
+  Evaluator<RHS1, Device> r1;
+  Evaluator<RHS2, Device> r2;
+
+  size_t nThr = 0;
+
+  Evaluator(Expression &expr)
+      : l(Evaluator<LHS, Device>(expr.l)),
+        r1(Evaluator<RHS1, Device>(expr.r1)),
+        r2(Evaluator<RHS2, Device>(expr.r2)) {
+    nThr = 2;
```

Comment:
There is something wrong because I can see the code ...  we will comment it on monday